File: ios/TappyAI/Features/UtilityTools/Model/FortuneEngine.swift
import Foundation

enum FortunePeriod: String, CaseIterable {
    case day, week, month

    var label: String {
        switch self {
        case .day: return "Hôm nay"
        case .week: return "Tuần này"
        case .month: return "Tháng này"
        }
    }
}

struct FortuneBanks {
    let love: [String]
    let career: [String]
    let money: [String]
    let health: [String]
    let luckyNumbers: [Int]
    let luckyColors: [String]
}

struct FortuneReading {
    let periodLabel: String
    let love: String
    let career: String
    let money: String
    let health: String
    let score: Int
    let luckyNumber: Int
    let luckyColor: String
}

enum FortuneEngine {
    static func getVNNow() -> Date {
        Date(timeIntervalSinceNow: 7 * 3600)
    }

    static func getPeriodKey(_ period: FortunePeriod) -> String {
        let vn = getVNNow()
        let cal = Calendar(identifier: .iso8601)
        let comps = cal.dateComponents(in: TimeZone(identifier: "Asia/Ho_Chi_Minh")!, from: vn)
        let y = comps.year!
        let m = comps.month!
        let d = comps.day!

        switch period {
        case .day:
            return "\(y)-\(String(format: "%02d", m))-\(String(format: "%02d", d))"
        case .week:
            let w = cal.component(.weekOfYear, from: vn)
            return "\(y)-W\(String(format: "%02d", w))"
        case .month:
            return "\(y)-\(String(format: "%02d", m))"
        }
    }

    static func hashString(_ s: String) -> Int {
        var h: Int32 = 5381
        for char in s.unicodeScalars {
            h = (h &* 33) &+ Int32(char.value)
        }
        return Int(abs(h))
    }

    static func pick<T>(_ seed: String, salt: String, from arr: [T]) -> T {
        let h = hashString("\(seed)|\(salt)")
        return arr[h % arr.count]
    }

    static func pickInRange(_ seed: String, salt: String, min: Int, max: Int) -> Int {
        let h = hashString("\(seed)|\(salt)")
        return min + (h % (max - min + 1))
    }

    static func generateFortune(subjectId: String, period: FortunePeriod, banks: FortuneBanks) -> FortuneReading {
        let key = getPeriodKey(period)
        let seed = "\(subjectId)|\(key)"

        return FortuneReading(
            periodLabel: period.label,
            love: pick(seed, salt: "love", from: banks.love),
            career: pick(seed, salt: "career", from: banks.career),
            money: pick(seed, salt: "money", from: banks.money),
            health: pick(seed, salt: "health", from: banks.health),
            score: pickInRange(seed, salt: "score", min: 3, max: 5),
            luckyNumber: pick(seed, salt: "number", from: banks.luckyNumbers),
            luckyColor: pick(seed, salt: "color", from: banks.luckyColors)
        )
    }

    static func generateYearFortune(subjectId: String, birthYear: Int, targetYear: Int) -> YearFortuneReading {
        let seed = "\(subjectId)|year-\(targetYear)"
        let subjectIdx = ((birthYear - 4) % 12 + 12) % 12
        let yearIdx = ((targetYear - 4) % 12 + 12) % 12
        let diff = (yearIdx - subjectIdx + 12) % 12
        let compat = YEAR_COMPAT[diff] ?? YearCompat(label: "—", note: "")

        return YearFortuneReading(
            periodLabel: "Năm \(targetYear) (\(YEAR_ANIMALS[yearIdx]))",
            love: pick(seed, salt: "love", from: YEAR_BANKS.love),
            career: pick(seed, salt: "career", from: YEAR_BANKS.career),
            money: pick(seed, salt: "money", from: YEAR_BANKS.money),
            health: pick(seed, salt: "health", from: YEAR_BANKS.health),
            score: pickInRange(seed, salt: "score", min: 3, max: 5),
            luckyNumber: pick(seed, salt: "number", from: YEAR_BANKS.luckyNumbers),
            luckyColor: pick(seed, salt: "color", from: YEAR_BANKS.luckyColors),
            yearAnimal: YEAR_ANIMALS[yearIdx],
            compatLabel: compat.label,
            compatNote: compat.note
        )
    }

    static func generateMonthlyBreakdown(subjectId: String, birthYear: Int, birthMonth: Int, birthDay: Int, targetYear: Int) -> [MonthlyFortune] {
        let vnMonths = ["Tháng 1", "Tháng 2", "Tháng 3", "Tháng 4", "Tháng 5", "Tháng 6", "Tháng 7", "Tháng 8", "Tháng 9", "Tháng 10", "Tháng 11", "Tháng 12"]
        return (0..<12).map { i in
            let month = i + 1
            let seed = "\(subjectId)|by\(birthYear)|bm\(birthMonth)|bd\(birthDay)|y\(targetYear)|m\(month)"
            return MonthlyFortune(
                month: month,
                monthName: vnMonths[i],
                love: pick(seed, salt: "love", from: MONTH_BANKS.love),
                career: pick(seed, salt: "career", from: MONTH_BANKS.career),
                money: pick(seed, salt: "money", from: MONTH_BANKS.money),
                health: pick(seed, salt: "health", from: MONTH_BANKS.health),
                note: pick(seed, salt: "note", from: MONTH_BANKS.note),
                score: pickInRange(seed, salt: "score", min: 2, max: 5)
            )
        }
    }

    static func generateLifeStages(subjectId: String, birthMonth: Int, birthDay: Int) -> [LifeStage] {
        STAGE_DEFS.map { def in
            guard let banks = STAGE_BANKS_DATA[def.key] else {
                return LifeStage(label: def.label, ageRange: def.ageRange, emoji: def.emoji, fate: "", career: "", love: "")
            }
            let seed = "\(subjectId)|bm\(birthMonth)|bd\(birthDay)|stage-\(def.key)"
            return LifeStage(
                label: def.label,
                ageRange: def.ageRange,
                emoji: def.emoji,
                fate: pick(seed, salt: "fate", from: banks.fate),
                career: pick(seed, salt: "career", from: banks.career),
                love: pick(seed, salt: "love", from: banks.love)
            )
        }
    }
}

struct YearFortuneReading {
    let periodLabel: String
    let love: String
    let career: String
    let money: String
    let health: String
    let score: Int
    let luckyNumber: Int
    let luckyColor: String
    let yearAnimal: String
    let compatLabel: String
    let compatNote: String
}

struct MonthlyFortune: Identifiable {
    let month: Int
    let monthName: String
    let love: String
    let career: String
    let money: String
    let health: String
    let note: String
    let score: Int

    var id: Int { month }
}

struct LifeStage: Identifiable {
    let label: String
    let ageRange: String
    let emoji: String
    let fate: String
    let career: String
    let love: String

    var id: String { label }
}

File: android/app/src/main/java/com/tappyai/app/fortune/engine/FortuneEngine.kt
package com.tappyai.app.fortune.engine

import java.time.LocalDate
import java.time.ZoneOffset
import java.time.temporal.WeekFields
import kotlin.math.abs

/**
 * Deterministic, fully-local fortune engine — a Kotlin port of the web `src/lib/boi/fortuneEngine.ts`
 * (prod worktree cool-vaughan-b3c7ff). No LLM, no network: a djb2 hash of `(subject + time bucket)`
 * picks a line from static string banks, so the same subject on the same day/week/month always yields
 * the same reading, but a new period yields a new one. Byte-identical to the web given the same banks
 * and seed ids.
 *
 * Two parity traps this port handles (see the web-vs-Kotlin notes on [hashString]):
 *  1. `| 0` 32-bit signed wrap — Kotlin `Int` arithmetic already wraps at 32 bits.
 *  2. `Math.abs` on `Int.MIN_VALUE` stays negative in Kotlin — widen to `Long` before abs.
 */

enum class FortunePeriod { Day, Week, Month }

data class FortuneBanks(
    val love: List<String>,
    val career: List<String>,
    val money: List<String>,
    val health: List<String>,
    val luckyNumbers: List<Int>,
    val luckyColors: List<String>,
)

data class FortuneReading(
    val periodLabel: String,
    val love: String,
    val career: String,
    val money: String,
    val health: String,
    val score: Int,        // 1-5 stars
    val luckyNumber: Int,
    val luckyColor: String,
)

data class PeriodKey(val key: String, val label: String)

/**
 * djb2 hash → non-negative Long. Web (`fortuneEngine.ts:62-69`):
 * `h = 5381; h = (h*33 + charCode) | 0; return Math.abs(h)`.
 * - Accumulate in `Int`: Kotlin Int arithmetic wraps at 32 bits exactly like JS `| 0`.
 * - Widen to `Long` before `abs` so `Int.MIN_VALUE` maps to `2147483648` (matching JS's double
 *   `Math.abs`), instead of `abs(Int.MIN_VALUE)` staying negative.
 * All seed inputs are pure ASCII, so `Char.code` == JS `charCodeAt`.
 */
internal fun hashString(s: String): Long {
    var h = 5381
    for (c in s) h = h * 33 + c.code
    return abs(h.toLong())
}

internal fun <T> pick(seed: String, salt: String, arr: List<T>): T {
    require(arr.isNotEmpty()) { "pick() needs a non-empty array" }
    val h = hashString("$seed|$salt")
    return arr[(h % arr.size).toInt()]
}

internal fun pickIndexInRange(seed: String, salt: String, min: Int, max: Int): Int {
    val h = hashString("$seed|$salt")
    return min + (h % (max - min + 1)).toInt()
}

private fun pad2(n: Int): String = n.toString().padStart(2, '0')

/**
 * The time-bucket key + VN label for a period, using Vietnam (UTC+7) civil date. Web parity trap
 * (`fortuneEngine.ts:44-60`): the week key is `"{calendarYear}-W{isoWeek}"` — the ISO-8601 week
 * NUMBER but the plain calendar year prefix (NOT the ISO week-based year). [vnDate] is injectable so
 * the key derivation is testable; production passes the real VN date.
 */
fun getPeriodKey(period: FortunePeriod, vnDate: LocalDate = LocalDate.now(ZoneOffset.ofHours(7))): PeriodKey {
    val y = vnDate.year
    return when (period) {
        FortunePeriod.Day -> PeriodKey("$y-${pad2(vnDate.monthValue)}-${pad2(vnDate.dayOfMonth)}", "Hôm nay")
        FortunePeriod.Week -> {
            val w = vnDate.get(WeekFields.ISO.weekOfWeekBasedYear())
            PeriodKey("$y-W${pad2(w)}", "Tuần này")
        }
        FortunePeriod.Month -> PeriodKey("$y-${pad2(vnDate.monthValue)}", "Tháng này")
    }
}

/**
 * Deterministic reading for one subject (sign / can-chi) in one period. Same subjectId + same period
 * bucket → same result. Mirrors `generateFortune` (`fortuneEngine.ts:86-100`). [vnDate] injectable for
 * tests. NOTE: [subjectId] must be the web string id (e.g. `aries`, `ty2`) for cross-platform parity.
 */
fun generateFortune(
    subjectId: String,
    period: FortunePeriod,
    banks: FortuneBanks,
    vnDate: LocalDate = LocalDate.now(ZoneOffset.ofHours(7)),
): FortuneReading {
    val (key, label) = getPeriodKey(period, vnDate)
    val seed = "$subjectId|$key"
    return FortuneReading(
        periodLabel = label,
        love = pick(seed, "love", banks.love),
        career = pick(seed, "career", banks.career),
        money = pick(seed, "money", banks.money),
        health = pick(seed, "health", banks.health),
        score = pickIndexInRange(seed, "score", 3, 5),
        luckyNumber = pick(seed, "number", banks.luckyNumbers),
        luckyColor = pick(seed, "color", banks.luckyColors),
    )
}
